datalake/coordinator: introduce table_commit_builder

nvartolomei opened the pull request:
Refactor the iceberg file committer so that it is _easier_ to extend to manage multiple tables (i.e. main table and the DLQ table).

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.3.x
- [ ] v24.2.x
- [ ] v24.1.x

## Release Notes
* none
<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

nvartolomei committed:
datalake: make execute_single_upload a function

Simpler code. Simpler to reuse from other functions.

nvartolomei committed:
datalake: extract upload_files method

Want to reuse it for DLQ files.

nvartolomei committed:
datalake/coordinator: refactor partition key building

commit_topic_files_to_catalog is too unwieldy. Simplify it before adding
DLQ logic.

nvartolomei committed:
datalake/coordinator: introduce table_commit_builder

Refactor the iceberg file committer so that it is _easier_ to extend to
manage multiple tables (i.e. main table and the DLQ table).

andrwng commented on src/v/datalake/translation_task.cc:
```diff
@@ -34,6 +34,29 @@ translation_task::errc map_error_code(cloud_data_io::errc errc) {
     }
 }
 
+ss::future<checked<remote_path, translation_task::errc>> execute_single_upload(
```
> datalake/coordinator: make execute_single_upload a function

nit: this isn't coordinator code. Prefer to just use datalake:

andrwng commented on src/v/datalake/translation_task.cc:
```diff
@@ -78,71 +78,23 @@ delete_local_data_files(
-  , _record_translator(&record_translator)
-  , _table_creator(&table_creator) {}
-
 ss::future<
```
>datalake/coordinator: extract upload_files method

Same here, just datalake, no coordinator

andrwng commented on src/v/datalake/coordinator/iceberg_file_committer.cc:
```diff
@@ -94,6 +94,106 @@ get_iceberg_committed_offset(const iceberg::table_metadata& table) {
 }
+
+checked<iceberg::struct_value, file_committer::errc> build_partition_key_struct(
+  const model::topic& topic,
```
Nice. Maybe it's worth using a prefix logger that includes the topic and table, if we expect more stuff to be reused by the DLT

nvartolomei replied:
I'll leave it as is for now.

andrwng commented on src/v/datalake/coordinator/iceberg_file_committer.cc:
```diff
@@ -194,6 +197,37 @@ checked<iceberg::partition_key, file_committer::errc> build_partition_key(
+  const iceberg::table_metadata& table,
+  const iceberg::manifest_io& io,
+  const chunked_vector<data_file>& files,
+  Out&& icb_files) {
```
Curious, why not just pass a chunked_vector<data_file>*?

nvartolomei replied:
What would be your motivation for a pointer here?

I somehow wanted to make it clear to the caller that some arguments are _"out"_ arguments and will be mutated. Using the output iterator made this very clear at the call site. Once I see the `std::back_inserter` it is clear what will happen.

_Would be nice to apply similar logic for the _"in"_ argument (`const chunked_vector<data_file>& files`) and use a const range instead but in C++ containers satisfy the range unlike in Rust (afaik) where you would have to explicitly call `C.iter()`_.

andrwng replied:
Pointer used to be a stylistic suggestion made by GSG for out-parameters (I think they've changed their stance here though). Of course, it has drawbacks when improperly used (e.g. people abusing "optional" out-parameters with nullptrs)

I guess an alternative is mutable reference, which may be a soft signal for out parameter, otherwise why else would it be mutable. I just see templates a pure overhead in reading this code if the main goal is to mark something as an out-parameter (could be a one-line comment even)

dotnwat replied:
`chunked_vector<data_file>*` makes sense to me. an l-value reference would also be ok i think, but i don't see how the forwarding reference is communicating intent.

andrwng commented on src/v/datalake/coordinator/iceberg_file_committer.cc:
```diff
@@ -259,6 +260,132 @@ ss::future<checked<std::nullopt_t, file_committer::errc>> commit_icb_files(
+    }
+
+private:
+    constexpr bool skip_entry(model::offset added_pending_at) const {
```
Curious why this is constexpr? Doesn't it rely on values that aren't known at compile time?

nvartolomei replied:
Yeah, probably not of much use.

andrwng commented on src/v/datalake/coordinator/iceberg_file_committer.cc:
```diff
@@ -259,6 +260,132 @@ ss::future<checked<std::nullopt_t, file_committer::errc>> commit_icb_files(
+/// topic, and will call process_pending_entry for each pending entry in the
+/// topic's state. Once all pending entries have been processed, the
+/// file_committer must call commit to commit the files to the Iceberg table.
+class table_committer {
```
nit: the "committer" name makes it easy to mistakenly think that this is meant to be long-running, considering the file committer is also nearby. How about `pending_table_commit` or something? IMO it feels clearer that it's a single commit's worth of accounting

nvartolomei replied:
`table_commit_builder` was my other choice

andrwng commented on src/v/datalake/coordinator/iceberg_file_committer.cc:
```diff
@@ -259,6 +260,132 @@ ss::future<checked<std::nullopt_t, file_committer::errc>> commit_icb_files(
+/// topic's state. Once all pending entries have been processed, the
+/// file_committer must call commit to commit the files to the Iceberg table.
+class table_committer {
+private:
```
nit: my personal preference is public methods up top, private methods below. Marking everything as public/private seems like a bit much (as clear as it is)... If the concern is findability of public/private markers, I'd also suggest moving the definitions out of the class declaration.

nvartolomei replied:
I prefer constructors to be at the top and in a separate public/private section and here I need them private. Does your preference still apply when talking about constructors?

nvartolomei replied:
https://google.github.io/styleguide/cppguide.html#Declaration_Order this is in line with your preference btw. I'll probably follow that.

andrwng replied:
Yea, I prefer separation of public and private, with private constructors being in a section below any public static constructors (basically what GSG says). If I'm reading a header (which I acknowledge this isn't) i'm usually probably trying to find public interface and it's helpful when it's all together

andrwng commented on src/v/datalake/coordinator/iceberg_file_committer.cc:
```diff
@@ -259,6 +260,132 @@ ss::future<checked<std::nullopt_t, file_committer::errc>> commit_icb_files(
+              table_id_,
+              table_commit_offset_);
+        } else {
+            auto res = collect_icb_files(
```
I'm not against keeping them, but I'm curious did the previous refactorings actually help reuse this code if we're pulling out a reusable table_committer abstraction?

nvartolomei replied:
It took a few iterations to arrive at the table committer abstractions. :)

Had to stop somewhere otherwise there is much more that can be improved. Given that I need to address the feedback here I'll try to do a couple more iterations tidying this code then.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--5873d5bf-31a1-43b5-b135-6c93dcc6ef6b-->
<!--TEST RESULTS-->
<!--0ca67fea-f8ff-4a3d-a65a-d16cf187b882-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60913">build#60913</a></summary>


| test_id | test_kind | job_url | test_status | passed |
|:--------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.partition_reassignments_test.PartitionReassignmentsTest.test_reassignments_kafka_cli | ducktape | https://buildkite.com/redpanda/redpanda/builds/60913#0194755a-de23-41d2-9408-03f0fbf40586 | FLAKY | 1/3 |
| rptest.tests.partition_reassignments_test.PartitionReassignmentsTest.test_reassignments_kafka_cli | ducktape | https://buildkite.com/redpanda/redpanda/builds/60913#0194755d-84d4-49c2-be7f-a2ecbfbf2d57 | FLAKY | 1/3 |
| test_archiver_manager_smp_rpfixture.test_archiver_manager_smp_rpfixture | unit | https://buildkite.com/redpanda/redpanda/builds/60913#01947516-78fc-4e70-88e4-639797670a65 | FLAKY | 1/2 |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/60960">build#60960</a></summary>


| test_id                                                                                                                                                                                  | test_kind   | job_url                                                                                   | test_status   | passed   |
|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|
| rptest.tests.audit_log_test.AuditLogTestsAppLifecycle.test_recovery_mode                                                                                                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/60960#019484dd-a61a-4329-acdf-27907d4a9910 | FLAKY         | 1/2      |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade                                                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/60960#019484c3-06ff-4531-9e40-05c4953e23f7 | FLAKY         | 1/2      |
| rptest.tests.compaction_recovery_test.CompactionRecoveryUpgradeTest.test_index_recovery_after_upgrade                                                                                    | ducktape    | https://buildkite.com/redpanda/redpanda/builds/60960#019484dd-a61b-4791-8dbb-624525b37148 | FLAKY         | 1/2      |
| rptest.tests.crl_test.CertificateRevocationTest.test_rpc                                                                                                                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/60960#019484c3-0700-4763-a218-d5a9b8b8ff25 | FAIL          | 0/20     |
| rptest.tests.partition_move_interruption_test.PartitionMoveInterruption.test_cancelling_partition_move.replication_factor=3.unclean_abort=True.recovery=restart_recovery.compacted=False | ducktape    | https://buildkite.com/redpanda/redpanda/builds/60960#019484c3-06ff-4531-9e40-05c4953e23f7 | FLAKY         | 1/2      |
</details>
<!--/body-->

dotnwat commented on src/v/datalake/coordinator/iceberg_file_committer.cc:
```diff
@@ -94,6 +98,294 @@ get_iceberg_committed_offset(const iceberg::table_metadata& table) {
+  iceberg::manifest_io& io,
+  iceberg::catalog& catalog,
+  const commit_offset_metadata commit_meta,
+  iceberg::table_metadata&& table,
+  chunked_vector<iceberg::data_file>&& icb_files) {
```
pass by value? it doesn't seem like there is much of a performance advantage here?

nvartolomei commented:
It’s not about forwarding reference but rather the output iterator concept.
The forwarding reference could be noise that I added unintentionally.

_As I refactored the code from chucked_vector& to output iterator version
it complained that lvalue is expected but std::back_inserter is giving me a
temporary so I changed that to Out&& from Out&._

On Sat, 18 Jan 2025 at 00:43, Noah Watkins ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/v/datalake/coordinator/iceberg_file_committer.cc
> <https://github.com/redpanda-data/redpanda/pull/24855#discussion_r1920908397>
> :
>
> > @@ -194,6 +197,37 @@ checked<iceberg::partition_key, file_committer::errc> build_partition_key(
>      return iceberg::partition_key{
>        std::make_unique<iceberg::struct_value>(std::move(pk_res.value()))};
>  }
> +
> +template<typename Out>
> +requires std::output_iterator<Out, iceberg::data_file>
> +checked<std::nullopt_t, file_committer::errc> collect_icb_files(
> +  const model::topic& topic,
> +  const iceberg::table_metadata& table,
> +  const iceberg::manifest_io& io,
> +  const chunked_vector<data_file>& files,
> +  Out&& icb_files) {
>
> chunked_vector<data_file>* makes sense to me. an l-value reference would
> also be ok i think, but i don't see how the forwarding reference is
> communicating intent.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/redpanda-data/redpanda/pull/24855#discussion_r1920908397>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEETWNTUB7FPI2GMAKHVAL2LGPSZAVCNFSM6AAAAABVMITHAKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDKNRQGIYTANZVGE>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/60960-->
<!--4fdb3d01-8b20-4efb-b57c-50d95e7ceae1-->
<!--title-->
## Retry command for [Build#60960](https://buildkite.com/redpanda/redpanda/builds/60960)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/crl_test.py::CertificateRevocationTest.test_rpc
```
<!--/body-->

nvartolomei commented:
/ci-repeat 1
tests/rptest/tests/crl_test.py::CertificateRevocationTest.test_rpc
